nixos/modules/config/sysctl.nix

```diff
@@ -59,5 +59,19 @@ in
     # Disable YAMA by default to allow easy debugging.
     boot.kernel.sysctl."kernel.yama.ptrace_scope" = mkDefault 0;
 
+    # Dynamic since kernel v5.12
+    # https://github.com/torvalds/linux/commit/ac7b79fd190b02e7151bc7d2b9da692f537657f3
+    boot.kernel.sysctl."fs.inotify.max_user_instances" =
+      if lib.versionAtLeast config.boot.kernelPackages.kernel.version "5.12"
+      then mkDefault 2147483647 # INT_MAX, dynamically limited based on available memory
+      else mkDefault 8192; # instead of 128
```

Comment:
> The max_user_instances sysctl limit defaults to 128 on NixOS and per user. When NixOS containers (nixos-container or declarative containers) are used, the inotify instances for root add up across containers, thus leading to exhaustion and Too many open files already for a few containers.

(https://github.com/NixOS/nixpkgs/issues/36214) &mdash; which we pretend fixing

Reply:
> (#36214) — which we pretend fixing

I don't quite understand what you mean -- but I think the anser to @alyssais's question is that the newer kernels mentioned in https://github.com/NixOS/nixpkgs/issues/36214#issuecomment-860249879 only set a dynamic kernel for `max_user_watches`, while this is `max_user_instances`, which is not dynamic.

That is, from the commit linked above:

> With inotify instances charged to memcg, the admin can simply set max_user_instances to INT_MAX and let the memcg limits of the jobs limit their inotify instances.

In other words, for `max_user_instances`, the linked commit instructs us to set the value to `INT_MAX` on newer kernels to be able to benefit from the automatic sizing (so that it isn't capped).

Does that answer your question @alyssais?

Reply:
... and in the case of older kernels, `128` might still not be enough for some use cases, so we default to `8192`, which does not come with a penalty unless those many instances are actually spawned (for reasons) on the user system.

Reply:
Ah, yes, I had misread the kernel commit.  I thought it would do that if no value was set, and setting a value would override it.  But we are indeed instructed to set it to INT_MAX.